requirements.txt

```diff
@@ -11,7 +11,7 @@ future>=0.15.2
 cryptography>=2.1.4
 pyparsing>=2.0.3,<2.4.0
 pyelftools>=0.22
-gdbgui>=0.13.2.0
+gdbgui<=0.13.2.1,>=0.13.2.0
 pygdbmi<=0.9.0.2
```

Comment:
gdbgui calls out which pygdbmi version it requires, so it shouldn't be included here. 

This project should consider using [pip-tools](https://github.com/jazzband/pip-tools#without-setuppy), which lets you specify things that should be installed at the top level, and then generates a requirements.txt lock file:

```
# requirements.in 
gdbgui==0.13.2.0
```

then compile the top level requirements into a lock file:

```
> pip-compile 
#
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile
#
brotli==1.0.9             # via flask-compress
click==7.1.2              # via flask
dnspython==2.0.0          # via eventlet
eventlet==0.25.2          # via gdbgui
flask-compress==1.5.0     # via gdbgui
flask-socketio==2.9.6     # via gdbgui
flask==0.12.5             # via flask-compress, flask-socketio, gdbgui
gdbgui==0.14.0.0          # via -r requirements.in
gevent-websocket==0.10.1  # via gdbgui
gevent==1.5.0             # via gdbgui, gevent-websocket
greenlet==0.4.16          # via eventlet, gevent
itsdangerous==1.1.0       # via flask
jinja2==2.11.2            # via flask
markupsafe==1.1.1         # via jinja2
monotonic==1.5            # via eventlet
pygdbmi==0.10.0.0         # via gdbgui
pygments==2.6.1           # via gdbgui
python-engineio==3.13.2   # via python-socketio
python-socketio==4.6.0    # via flask-socketio
six==1.15.0               # via eventlet, python-socketio
werkzeug==0.16.1          # via flask
```

It also includes a dependency resolver, which ensures there are no conflicting requirements in the installed versions.

Reply:
Hi @cs01, pygdbmi is also a "top-level" requirement in ESP-IDF, this is why it is mentioned here explicitly. Thanks for the pointer about pip-tools, it looks very useful!